llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp

```diff
@@ -1504,17 +1504,32 @@ Instruction *InstCombinerImpl::visitLShr(BinaryOperator &I) {
       // able to invert the transform and perf may suffer with an extra mul
       // instruction.
       if (Op0->hasOneUse()) {
-        APInt NewMulC = MulC->lshr(ShAmtC);
-        // if c is divisible by (1 << ShAmtC):
-        // lshr (mul nuw x, MulC), ShAmtC -> mul nuw nsw x, (MulC >> ShAmtC)
-        if (MulC->eq(NewMulC.shl(ShAmtC))) {
-          auto *NewMul =
-              BinaryOperator::CreateNUWMul(X, ConstantInt::get(Ty, NewMulC));
-          assert(ShAmtC != 0 &&
-                 "lshr X, 0 should be handled by simplifyLShrInst.");
-          NewMul->setHasNoSignedWrap(true);
-          return NewMul;
+        unsigned CommonZeros = std::min(MulC->countr_zero(), ShAmtC);
+        if (CommonZeros != 0) {
+          APInt NewMulC = MulC->lshr(CommonZeros);
+          unsigned NewShAmtC = ShAmtC - CommonZeros;
+          // if c is divisible by (1 << ShAmtC):
+          // lshr (mul nuw x, MulC), ShAmtC -> mul nuw nsw x, (MulC >> ShAmtC)
+          if (NewShAmtC == 0) {
+            auto *NewMul =
+                BinaryOperator::CreateNUWMul(X, ConstantInt::get(Ty, NewMulC));
+            NewMul->setHasNoSignedWrap(true);
+            return NewMul;
+          }
+
+          // We can reduce expressions such as like lshr (mul nuw x, 6), 2 ->
+          // lshr (mul nuw nsw x, 3), 1
+          // TODO: What about if ALL uses can be simplified in this way? Is that
+          // likely enough to happen to justify even caring?
+          auto *NewMul = Builder.CreateMul(X, ConstantInt::get(Ty, NewMulC), "",
+                                           /*NUW*/ true, /*NSW*/ true);
```

Comment:
Style on the constants, `/*NUW=*/true, /*NSW=*/true`.

As an aside, whats the intuition for why its okay to add `nsw` here?

Reply:
In simple terms:

Consider x * c. Result's most significant bit can either only be one or zero. For this transformation to be applied, c must be divisible by 2. Given how nuw means that no set bits were lost in the unsigned multiplication, it therefore follows that (x * c) is equal to (x * c) / 2  * 2, which is exact because c is divisible by 2. 

For (x * c) / 2 * 2 to be nuw, the most significant bit of (x * c) / 2 must be 0, so that x * c doesn't overflow/shift out a 1.

Therefore, x * c must be a positive number that is less than the signed max, and therefore x * c is nsw.

Reply:
Also see alive2 proof